pkg/data/implementations/gcp_remote_url.go

```diff
@@ -0,0 +1,189 @@
+
+	credentials "cloud.google.com/go/iam/credentials/apiv1"
+	gcs "cloud.google.com/go/storage"
+	"github.com/lyft/flyteadmin/pkg/data/interfaces"
+	"github.com/lyft/flyteadmin/pkg/errors"
+	"github.com/lyft/flytestdlib/logger"
+	"github.com/lyft/flytestdlib/storage"
+	"google.golang.org/api/option"
+	credentialspb "google.golang.org/genproto/googleapis/iam/credentials/v1"
+	"google.golang.org/grpc/codes"
+)
+
+const gcsScheme = "gs"
+
+type iamCredentialsInterface interface {
+	SignBlob(ctx context.Context, req *credentialspb.SignBlobRequest, opts ...gax.CallOption) (*credentialspb.SignBlobResponse, error)
+	GenerateAccessToken(ctx context.Context, req *credentialspb.GenerateAccessTokenRequest, opts ...gax.CallOption) (*credentialspb.GenerateAccessTokenResponse, error)
+}
+
+type gcsClientWrapper struct {
+	delegate *gcs.Client
+}
+
+type bucketHandleWrapper struct {
+	delegate *gcs.BucketHandle
+}
+
+type objectHandleWrapper struct {
+	delegate *gcs.ObjectHandle
+}
+
+type gcsInterface interface {
+	Bucket(name string) bucketHandleInterface
+}
+
+type bucketHandleInterface interface {
+	Object(name string) objectHandleInterface
+}
+
+type objectHandleInterface interface {
+	Attrs(ctx context.Context) (attrs *gcs.ObjectAttrs, err error)
+}
+
+// GCP-specific implementation of RemoteURLInterface
+type GCPRemoteURL struct {
+	iamCredentialsClient iamCredentialsInterface
+	gcsClient            gcsInterface
+	signDuration         time.Duration
+	signingPrincipal     string
+}
+
+type GCPGCSObject struct {
+	bucket string
+	object string
+}
+
+type impersonationTokenSource struct {
+	iamCredentialsClient iamCredentialsInterface
+	signingPrincipal     string
+}
+
+func (c *gcsClientWrapper) Bucket(name string) bucketHandleInterface {
+	return &bucketHandleWrapper{delegate: c.delegate.Bucket(name)}
+}
+
+func (b *bucketHandleWrapper) Object(name string) objectHandleInterface {
+	return &objectHandleWrapper{delegate: b.delegate.Object(name)}
+}
+
+func (o *objectHandleWrapper) Attrs(ctx context.Context) (attrs *gcs.ObjectAttrs, err error) {
+	return o.delegate.Attrs(ctx)
+}
+
+func (g *GCPRemoteURL) splitURI(ctx context.Context, uri string) (GCPGCSObject, error) {
+	scheme, container, key, err := storage.DataReference(uri).Split()
+	if err != nil {
+		return GCPGCSObject{}, err
+	}
+	if scheme != gcsScheme {
+		logger.Debugf(ctx, "encountered unexpected scheme: %s for GCS URI: %s", scheme, uri)
+		return GCPGCSObject{}, errors.NewFlyteAdminErrorf(codes.InvalidArgument,
+			"unexpected scheme %s for GCS URI", scheme)
+	}
+	return GCPGCSObject{
+		bucket: container,
+		object: key,
+	}, nil
+}
+
+func (g *GCPRemoteURL) signURL(ctx context.Context, gcsURI GCPGCSObject) (string, error) {
+	opts := &gcs.SignedURLOptions{
+		Method:         "GET",
+		GoogleAccessID: g.signingPrincipal,
+		SignBytes: func(b []byte) ([]byte, error) {
+			req := &credentialspb.SignBlobRequest{
+				Payload: b,
+				Name:    "projects/-/serviceAccounts/" + g.signingPrincipal,
```

Comment:
what's the meaning of this string prefix?

Reply:
It's a GCP resource name: https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/signBlob

Basically, it means the `service account` in some GCP project (`-`) that doesn't matter.